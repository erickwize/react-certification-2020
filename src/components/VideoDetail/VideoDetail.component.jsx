import React from 'react';
import { useParams } from 'react-router-dom';
import { VideoFrame }from './VideoFrame/VideoFrame.component';
import { VideoInfo }from './VideoInfo/VideoInfo.component';
import { RelatedVideos }from './RelatedVideos/RelatedVideos.component';
import { Container, Subcontainer, LeftContainer, RightContainer} from './VideoDetail.styles';
import { useFetchVideoInfo } from '../../utils/hooks/useFetchVideoInfo';

export const VideoDetail = () => {
    const { videoId } = useParams();
    const { videos } = useFetchVideoInfo(videoId);
    const video = videos.items[0];

    return <Container data-testid="video-detail">
        <Subcontainer>
            <LeftContainer>
                <VideoFrame videoId={videoId}></VideoFrame>
                {video && <VideoInfo video={video}></VideoInfo>}
            </LeftContainer>
            <RightContainer>
                <RelatedVideos videoId={videoId}></RelatedVideos>
            </RightContainer>
        </Subcontainer>
    </Container>
}