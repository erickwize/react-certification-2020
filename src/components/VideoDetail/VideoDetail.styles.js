import styled from 'styled-components';

export const Container = styled.div`
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
`;

export const Subcontainer = styled.div`
    width: 95%;
`;

export const LeftContainer = styled.div`
    float: left;
    width: 75%;

    @media(max-width: 1024px){
        width: 100%;
    }
`;

export const RightContainer = styled.div`
    float: left;
    width: 25%;

    @media(max-width: 1024px){
        margin-top: 1rem;
        width: 100%;
    }
`;