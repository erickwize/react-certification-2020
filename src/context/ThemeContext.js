import React, { createContext, useState } from 'react'

const ThemeContext = createContext()

const THEMES = {
	light: {
		background: '#eeeeee',
		color: '#555555',
		colorTitle: '#333333',
		headerBackground: '#dddddd',
		cardBackground: '#e6e6e6',
	},
	dark: {
		background: '#33333d',
		color: '#b7b7ba',
		colorTitle: '#f9f9f9',
		headerBackground: '#424251',
		cardBackground: '#1a1a1a',
	},
}
function ThemeContextProvider({ children }) {
	const [theme, setTheme] = useState('dark')
	const defaultContext = {
		theme,
		toggle: () => setTheme(theme === 'light' ? 'dark' : 'light'),
		...THEMES[theme],
	}

	return <ThemeContext.Provider value={defaultContext}>{children}</ThemeContext.Provider>
}

export { ThemeContext, ThemeContextProvider }
