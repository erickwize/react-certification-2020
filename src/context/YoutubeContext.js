import React, { createContext, useState } from 'react'
import { saveArrayOnLocalStorage } from '../utils/localStorage'

const API_KEY = process.env.REACT_APP_YT_API_KEY
const MAX_RESULTS = 20

const YoutubeContext = createContext()

const SEARCH_URL = `https://www.googleapis.com/youtube/v3/search?part=snippet&key=${API_KEY}&type=video&maxResults=${MAX_RESULTS}&`

function YoutubeContextProvider({ children }) {
	const [videos, setVideos] = useState([])
	const defaultContext = {
		videos,
		search: term => {
			fetch(`${SEARCH_URL}q=${term}`)
				.then(response => response.json())
				.then(data => {
					if (data?.items?.length) {
						setVideos(data.items)
						saveArrayOnLocalStorage('videos', data.items)
					} else setVideos([])
				})
				.catch(error => {
					console.error(error)
				})
		},
		setVideosContext: newVideos => {
			setVideos(newVideos)
		},
	}

	return (
		<YoutubeContext.Provider value={defaultContext}>{children}</YoutubeContext.Provider>
	)
}

export { YoutubeContext, YoutubeContextProvider }
