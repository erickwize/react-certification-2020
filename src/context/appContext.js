import { createContext, useContext } from "react";

const appContext = createContext();

export const useAppContext = () => {
  const context = useContext(appContext);
  if (context === undefined) {
    throw new Error(
      "useAppContext must be used within AppState component tree"
    );
  }
  return context;
};

export default appContext;
