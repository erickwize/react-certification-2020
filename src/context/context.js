// import Dexie from 'dexie'
import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'

export const Context = createContext(null)

export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    default:
      return state
  }
}

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
