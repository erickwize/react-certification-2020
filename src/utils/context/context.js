import { createContext } from 'react';

export const Context = createContext({
  queryToSearch: null,
  setQueryFromSearchInput: null,
});
