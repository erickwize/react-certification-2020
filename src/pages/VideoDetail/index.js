import React, { useContext } from 'react'
import styled from 'styled-components'
import { v4 as uuidv4 } from 'uuid'
import { Link } from 'react-router-dom'
import { ThemeContext } from '../../context/ThemeContext'
import useVideoDetail from '../../customHooks/useVideoDetail'

const VideoDetailStyled = styled.div`
	padding: 20px 80px;
	display: flex;
`
const VideoInfoStyled = styled.div`
	margin-right: 20px;
	color: ${props => props.theme.color};
`
const RelatedVideosStyled = styled.div`
	a {
		text-decoration: inherit;
	}
`
const ThumbnailVideoStyled = styled.div`
	display: grid;
	grid-template-columns: 120px auto;
	margin-bottom: 8px;
	background: ${props => props.theme.headerBackground};
	p {
		margin-top: 8px;
		font-size: 12px;
	}
	span {
		font-size: 10px;
	}
`
const ThumbnailVideoDescriptionStyled = styled.div`
	padding: 0px 8px;
`
const IframeStyled = styled.iframe`
	width: 900px;
	height: 500px;
	border: 0;
`

export default function VideoDetail() {
	const themeContext = useContext(ThemeContext)
	const [currentVideo, relatedVideos, videoId] = useVideoDetail()

	const {
		snippet: { title, channelTitle, description },
	} = currentVideo

	return (
		<VideoDetailStyled>
			<VideoInfoStyled theme={themeContext}>
				<IframeStyled title={title} src={`https://www.youtube.com/embed/${videoId}`} />
				<h2>{title}</h2>
				<h4>{channelTitle}</h4>
				<p>{description}</p>
			</VideoInfoStyled>
			<RelatedVideosStyled>
				{relatedVideos.map(video => {
					return (
						<Link key={uuidv4()} to={`/video-detail/${video.id.videoId}`}>
							<ThumbnailVideoStyled theme={themeContext}>
								<img
									src={video.snippet.thumbnails.default.url}
									alt={video.snippet.title}
								/>
								<ThumbnailVideoDescriptionStyled>
									<p>{video.snippet.title}</p>
									<span>{video.snippet.channelTitle}</span>
								</ThumbnailVideoDescriptionStyled>
							</ThumbnailVideoStyled>
						</Link>
					)
				})}
			</RelatedVideosStyled>
		</VideoDetailStyled>
	)
}
