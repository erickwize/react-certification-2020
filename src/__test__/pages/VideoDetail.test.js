import { render } from '@testing-library/react'
import { Route, Router } from 'react-router'
import { createMemoryHistory } from 'history'
import React from 'react'
import { ThemeContextProvider } from '../../context/ThemeContext'
import { YoutubeContextProvider } from '../../context/YoutubeContext'
import VideoDetail from '../../pages/VideoDetail'

const history = createMemoryHistory()
history.push('/video-detail/testId123456')

jest.mock('react-router', () => ({
	...jest.requireActual('react-router'),
	useParams: () => ({
		videoId: 'testId123456',
	}),
}))

export const renderWithRouter = Component =>
	render(
		<YoutubeContextProvider>
			<ThemeContextProvider>
				<Router history={history}>
					<Route>
						<Component />
					</Route>
				</Router>
			</ThemeContextProvider>
		</YoutubeContextProvider>
	)

describe('VideoDetail testing', () => {
	it('Video Detail Snapshot', () => {
		const { container } = renderWithRouter(VideoDetail)
		expect(container).toMatchSnapshot()
	})
})
