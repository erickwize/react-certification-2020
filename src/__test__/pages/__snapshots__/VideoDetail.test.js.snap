// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetail testing Video Detail Snapshot 1`] = `
<div>
  <div
    class="sc-bdfBwQ duyacE"
  >
    <div
      class="sc-gsTCUz iiRrAN"
    >
      <iframe
        class="sc-jSgupP dVHUwC"
        src="https://www.youtube.com/embed/testId123456"
        title="..."
      />
      <h2>
        ...
      </h2>
      <h4 />
      <p />
    </div>
    <div
      class="sc-dlfnbm iLwtyW"
    />
  </div>
</div>
`;
